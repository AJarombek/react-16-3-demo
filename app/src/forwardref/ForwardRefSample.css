/**
 * Stylesheet for the ForwardRefSample components.
 * Author: Andrew Jarombek
 * Date: 1/16/2020
 */

.forward-ref-sample {
    display: flex;
    align-items: center;
    justify-content: center;
}

.prop-ref, .forward-ref {
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center;
    margin: 0 20px;
}
