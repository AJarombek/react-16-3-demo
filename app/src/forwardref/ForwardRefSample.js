/**
 * Demonstrate how forward refs work.
 * @author Andrew Jarombek
 * @since 1/15/2020
 */

import React, {createRef, useState} from 'react';
import ButtonWrapper from './ButtonWrapper';
import ForwardRefButtonWrapper from './ForwardRefButtonWrapper';
import { AJCodeSnippet } from 'jarombek-react-components';

const ForwardRefSample = () => {
  // Create refs to attach to the custom button wrapper components.
  const buttonWrapperRef = createRef();
  const buttonWrapperForwardRef = createRef();

  // Create hooks which are used to display (render) the value of the refs.
  const [propRef, setPropRef] = useState("");
  const [forwardRef, setForwardRef] = useState("");

  // Action which occurs when the ButtonWrapper component is clicked.
  const buttonWrapperOnClick = () => {
    console.info(buttonWrapperRef.current);
    setPropRef(buttonWrapperRef.current);
  };

  // Action which occurs when the ButtonWrapperForwardRef component is clicked.
  const buttonWrapperForwardRefOnClick = () => {
    console.info(buttonWrapperForwardRef.current);
    buttonWrapperForwardRef.current = buttonWrapperForwardRef.current.toString();
    setForwardRef(buttonWrapperForwardRef);
  };

  return (
    <div className="forward-ref-sample">
      <div className="prop-ref">
        <h3>Ref on Component WITHOUT a Forward Ref</h3>
        <ButtonWrapper
          onClick={() => buttonWrapperOnClick()}
          ref={buttonWrapperRef}>

          No Forward Ref
        </ButtonWrapper>
        <p>{JSON.stringify(propRef)}</p>
      </div>
      <div className="forward-ref">
        <h3>Ref on Component WITH a Forward Ref</h3>
        <ForwardRefButtonWrapper
          onClick={() => buttonWrapperForwardRefOnClick()}
          ref={buttonWrapperForwardRef}>

          Forward Ref
        </ForwardRefButtonWrapper>
        <p>{JSON.stringify(forwardRef)}</p>
      </div>
    </div>
  );
};

export default ForwardRefSample;
