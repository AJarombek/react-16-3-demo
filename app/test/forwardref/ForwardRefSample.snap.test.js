/**
 * Snapshot test for the {@link ../../src/forwardref/ForwardRefSample} component.
 * @author Andrew Jarombek
 * @since 1/28/2020
 */

import React from 'react';
import ForwardRefSample from '../../src/forwardref/ForwardRefSample';
import renderer from 'react-test-renderer';

it('renders correctly', () => {
  const tree = renderer.create(<ForwardRefSample />).toJSON();
  expect(tree).toMatchSnapshot();
});
