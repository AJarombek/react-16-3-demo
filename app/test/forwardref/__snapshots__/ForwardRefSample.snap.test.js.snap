// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="forward-ref-sample"
>
  <div
    className="prop-ref"
  >
    <h3>
      Ref on Component WITHOUT a Forward Ref
    </h3>
    <button
      className="button-wrapper"
      onClick={[Function]}
      type="button"
    >
      No Forward Ref
    </button>
    <p>
      ""
    </p>
  </div>
  <div
    className="forward-ref"
  >
    <h3>
      Ref on Component WITH a Forward Ref
    </h3>
    <button
      className="button-wrapper"
      onClick={[Function]}
      type="button"
    >
      Forward Ref
    </button>
    <p>
      ""
    </p>
  </div>
</div>
`;
